netket/optimizer/stochastic_reconfiguration.py

```diff
@@ -0,0 +1,303 @@
+import numpy as _np
+from scipy.linalg import lstsq as _lstsq
+from scipy.sparse.linalg import LinearOperator
+from netket.stats import sum_on_nodes as _sum_on_nodes
+from scipy.sparse.linalg import cg, gmres, minres
+from mpi4py import MPI
+
+
+class SR:
+    r"""
+    Performs stochastic reconfiguration (SR) updates.
+    """
+
+    def __init__(
+        self,
+        lsq_solver=None,
+        diag_shift=0.01,
+        use_iterative=True,
```

Comment:
`use_iterative` is now determined solely by the `lsq_solver` option as it is overwritten below in any case. So it should probably not be a separate parameter.

Reply:
Previously, the default was to not use an iterative solver. Is there a reason for changing this default?

Reply:
Perhaps it might be good to actually keep the `use_iterative` flag for the moment. I think sooner or later I shall be trying to make a PR on using the cholesky solver, which will likely imply a `use_cholesky` flag (as we did previously).

Reply:
> `use_iterative` is now determined solely by the `lsq_solver` option as it is overwritten below in any case. So it should probably not be a separate parameter.

there is the case when lsq_solver=None (default value) and use_iterative=True. In that case use_iterative provided in the constructor is used... Basically the most common use pattern would be to call this without specifying lsq_solver but just use_iterative=True/False.

Reply:
> Previously, the default was to not use an iterative solver. Is there a reason for changing this default?

I noticed that the sparse solver is faster anyways, even on small systems... but I didn't test this extensively, what's your feeling here?

Reply:
> Perhaps it might be good to actually keep the `use_iterative` flag for the moment. I think sooner or later I shall be trying to make a PR on using the cholesky solver, which will likely imply a `use_cholesky` flag (as we did previously).

@fabienalet yes it makes sense to keep the use_iterative flag (see other comments!) For the cholesky, I would suggest to implement it as an optional ```lsq_solver='cholesky'``` in the constructor of SR or something like that, without introducing a new flag

Reply:
OK, well noted.

Reply:
Trying to implement this, but I see the cholesky solver is very sensible to the shift and easily run into instabilities. I'm trying to solve this by comparing systematically to the QR solver. Quick basic question: what is the way to fix all random number generators to be the same in two different runs? I thought fixing the seed of the machine, as well as netket.random.seed would do it but it doesn't seem to be the case as(maybe I'm missing something obvious/did a mistake)

Reply:
@fabienalet yes if you are using the c++ sampler backend for the moment you need to fix also ```netket.utils.seed``` ... this is going to be removed though